File: Chapter 03/Python/section_3_2_coupling_exercise/after/user_report_generator.py
from .user_data_service import UserDataService

"""
THE SOLUTION: The Loosely Coupled Client.
* ARCHITECTURE NOTE: Look at how much simpler this generator has become. 
It no longer contains a loop, it doesn't perform any math, and it doesn't 
need to know about 'orders'. 
* It simply asks for a 'report' and formats it. This client is now 
"Ignorant by Design"—it only knows about the UserReportData contract, 
not the underlying data source or logic.
"""
class UserReportGenerator:
    def __init__(self):
        self.data_service = UserDataService()

    def generate_report(self, user_id: int) -> str:
        # A single call replaces the "Chatty" sequential calls.
        report = self.data_service.get_user_report(user_id)
        
        return f"User Report for {report.name} ({report.email}) - Total Spent: ${report.total_spent:.2f}"

File: Chapter 03/JS/section_3_2_coupling_exercise/after/userReportGenerator.js
const UserDataService = require('./userDataService');

/**
 * THE SOLUTION: The Loosely Coupled Client.
 * * ARCHITECTURE NOTE: Look at how clean and simple the client has become.
 * It no longer suffers from "Feature Envy"—it doesn't have to loop through 
 * orders, request individual totals, or perform its own math.
 * * It simply asks the service for the completed report data and formats it. 
 * If the underlying database schema or the way a "totalSpent" is calculated 
 * changes, this client class will not require a single line of code to be modified.
 */
class UserReportGenerator {
    constructor() {
        this.dataService = new UserDataService();
    }

    /**
     * Generates a formatted string report for a user using a single service call.
     * @param {number} userId - The ID of the user.
     * @returns {string} The formatted report.
     */
    generateReport(userId) {
        // A single, clean call replaces multiple chatty calls.
        const report = this.dataService.getUserReport(userId);
        
        return `User Report for ${report.name} (${report.email}) - Total Spent: $${report.totalSpent.toFixed(2)}`;
    }
}

module.exports = UserReportGenerator;